runtime/major_gc.c

```diff
@@ -316,13 +318,26 @@ static void ephe_todo_list_emptied (void)
 
   /* Since the todo list is empty, this domain does not need to participate in
    * further ephemeron cycles. */
-  atomic_fetch_sub(&ephe_cycle_info.num_domains_todo, 1);
+  atomic_fetch_add_verify_ge0(&ephe_cycle_info.num_domains_todo, -1);
   CAMLassert(atomic_load_acquire(&ephe_cycle_info.num_domains_done) <=
              atomic_load_acquire(&ephe_cycle_info.num_domains_todo));
 
   caml_plat_unlock(&ephe_lock);
 }
 
+/* Begin ephemeron marking by making all 'live' ephes become 'todo' */
+static void begin_ephe_marking(void)
```

Comment:
In trunk, newly allocated ephemerons go into the live list since they are allocated black. As a result, these ephemerons don't need to be marked in the current cycle. With this change, all the live ephemerons are brought into the todo list at the start of the ephemeron marking, which is correct but less efficient than the current scheme. 

As an alternative to this, would it not be sufficient to check whether marking has started in [`caml_ephe_create`](https://github.com/ocaml/ocaml/blob/f5ff74224029dc9ed765fe0942bd7fd41b194e39/runtime/weak.c#L60). If we marking has not started, add the ephemeron to the todo list. Otherwise, add it to the live list.

Reply:
I re-read the code again, and the change I proposed is equivalent to the design here. In particular, I had assumed, based on the name of the function, that this function was called before `ephe_mark`. In fact, this is called at the beginning of `Phase_sweep_and_mark_main`.

Reply:
I've renamed the function.